File: app/src/main/java/com/kickstarter/viewmodels/MessageThreadsViewModel.java
package com.kickstarter.viewmodels;

import android.graphics.Typeface;
import android.util.Pair;

import com.kickstarter.R;
import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.ApiPaginator;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.KoalaContext;
import com.kickstarter.libs.RefTag;
import com.kickstarter.libs.utils.IntegerUtils;
import com.kickstarter.libs.utils.ObjectUtils;
import com.kickstarter.libs.utils.PairUtils;
import com.kickstarter.models.MessageThread;
import com.kickstarter.models.Project;
import com.kickstarter.models.User;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.apiresponses.MessageThreadsEnvelope;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.MessageThreadsActivity;
import com.kickstarter.ui.data.Mailbox;
import com.kickstarter.ui.intentmappers.ProjectIntentMapper;

import java.util.List;

import androidx.annotation.NonNull;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.combineLatestPair;
import static com.kickstarter.libs.rx.transformers.Transformers.ignoreValues;
import static com.kickstarter.libs.rx.transformers.Transformers.neverError;
import static com.kickstarter.libs.rx.transformers.Transformers.takeWhen;
import static com.kickstarter.libs.utils.IntegerUtils.intValueOrZero;

public interface MessageThreadsViewModel {

  interface Inputs {
    void mailbox(Mailbox mailbox);

    /** Call when pagination should happen. */
    void nextPage();

    /** Call when onResume of the activity's lifecycle happens. */
    void onResume();

    /** Call when the swipe refresher is invoked. */
    void swipeRefresh();
  }

  interface Outputs {
    /** Emits a boolean to determine if there are no messages. */
    Observable<Boolean> hasNoMessages();

    /** Emits a boolean to determine if there are no unread messages. */
    Observable<Boolean> hasNoUnreadMessages();

    /** Emits a boolean indicating whether message threads are being fetched from the API. */
    Observable<Boolean> isFetchingMessageThreads();

    /** Emits a string resource integer to set the mailbox title text view to. */
    Observable<Integer> mailboxTitle();

    /** Emits a list of message threads to be displayed. */
    Observable<List<MessageThread>> messageThreadList();

    /** Emits a color integer to set the unread count text view to. */
    Observable<Integer> unreadCountTextViewColorInt();

    /** Emits a typeface integer to set the unread count text view to. */
    Observable<Integer> unreadCountTextViewTypefaceInt();

    /** Emits a boolean to determine if the unread count toolbar text view should be gone. */
    Observable<Boolean> unreadCountToolbarTextViewIsGone();

    /** Emits the unread message count to be displayed. */
    Observable<Integer> unreadMessagesCount();

    /** Emits a boolean determining the unread messages count visibility. */
    Observable<Boolean> unreadMessagesCountIsGone();
  }

  final class ViewModel extends ActivityViewModel<MessageThreadsActivity> implements Inputs, Outputs {
    private final ApiClientType client;
    private final CurrentUserType currentUser;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.client = environment.apiClient();
      this.currentUser = environment.currentUser();

      // NB: project from intent can be null.
      final Observable<Project> initialProject = intent()
        .map(i -> i.getParcelableExtra(IntentKey.PROJECT));

      final Observable<KoalaContext.Mailbox> koalaContext = intent()
        .map(i -> i.getSerializableExtra(IntentKey.KOALA_CONTEXT))
        .ofType(KoalaContext.Mailbox.class);

      final Observable<Void> refreshUserOrProject = Observable.merge(this.onResume, this.swipeRefresh);

      final Observable<User> freshUser = intent()
        .compose(takeWhen(refreshUserOrProject))
        .switchMap(__ -> this.client.fetchCurrentUser())
        .retry(2)
        .compose(neverError());

      freshUser.subscribe(this.currentUser::refresh);

      final Observable<Project> project = Observable.merge(
        initialProject,
        initialProject
          .compose(takeWhen(refreshUserOrProject))
          .map(Project::param)
          .switchMap(this.client::fetchProject)
          .compose(neverError())
          .share()
      );

      // Use project unread messages count if configured with a project,
      // in the case of a creator viewing their project's messages.
      final Observable<Integer> unreadMessagesCount = Observable.combineLatest(
        project,
        this.currentUser.loggedInUser(),
        Pair::create
      )
        .map(projectAndUser ->
          projectAndUser.first != null
            ? projectAndUser.first.unreadMessagesCount()
            : projectAndUser.second.unreadMessagesCount()
        )
        .distinctUntilChanged();

      // todo: MessageSubject switch will also trigger refresh
      final Observable<Void> refreshMessageThreads = Observable.merge(
        unreadMessagesCount.compose(ignoreValues()),
        this.swipeRefresh
      );

      final Observable<Mailbox> mailbox = this.mailbox
        .startWith(Mailbox.INBOX)
        .distinctUntilChanged();

      mailbox
        .map(this::getStringResForMailbox)
        .compose(bindToLifecycle())
        .subscribe(this.mailboxTitle);

      final Observable<Pair<Project, Mailbox>> projectAndMailbox = Observable.combineLatest(
        project.distinctUntilChanged(), mailbox.distinctUntilChanged(), Pair::create);

      final Observable<Pair<Project, Mailbox>> startOverWith = Observable.combineLatest(
        projectAndMailbox,
        refreshMessageThreads,
        Pair::create
      )
        .map(PairUtils::first);

      final ApiPaginator<MessageThread, MessageThreadsEnvelope, Pair<Project, Mailbox>> paginator =
        ApiPaginator.<MessageThread, MessageThreadsEnvelope, Pair<Project, Mailbox>>builder()
          .nextPage(this.nextPage)
          .startOverWith(startOverWith)
          .envelopeToListOfData(MessageThreadsEnvelope::messageThreads)
          .envelopeToMoreUrl(env -> env.urls().api().moreMessageThreads())
          .loadWithParams(pm -> this.client.fetchMessageThreads(pm.first, pm.second))
          .loadWithPaginationPath(this.client::fetchMessageThreadsWithPaginationPath)
          .clearWhenStartingOver(true)
          .build();

      paginator.isFetching()
        .compose(bindToLifecycle())
        .subscribe(this.isFetchingMessageThreads);

      paginator.paginatedData()
        .distinctUntilChanged()
        .compose(bindToLifecycle())
        .subscribe(this.messageThreadList);

      unreadMessagesCount
        .map(ObjectUtils::isNull)
        .subscribe(this.hasNoMessages);

      unreadMessagesCount
        .map(IntegerUtils::isZero)
        .subscribe(this.hasNoUnreadMessages);

      unreadMessagesCount
        .map(count -> intValueOrZero(count) > 0 ? R.color.accent : R.color.ksr_dark_grey_400)
        .subscribe(this.unreadCountTextViewColorInt);

      unreadMessagesCount
        .map(count -> intValueOrZero(count) > 0 ? Typeface.BOLD : Typeface.NORMAL)
        .subscribe(this.unreadCountTextViewTypefaceInt);

      this.unreadCountToolbarTextViewIsGone = Observable.zip(
        this.hasNoMessages,
        this.hasNoUnreadMessages,
        Pair::create
      )
        .map(noMessagesAndNoUnread -> noMessagesAndNoUnread.first || noMessagesAndNoUnread.second)
        .compose(combineLatestPair(mailbox))
        .map(noMessagesAndMailbox -> noMessagesAndMailbox.first || noMessagesAndMailbox.second.equals(Mailbox.SENT));

      unreadMessagesCount
        .filter(ObjectUtils::isNotNull)
        .filter(IntegerUtils::isNonZero)
        .subscribe(this.unreadMessagesCount);

      this.unreadMessagesCountIsGone = mailbox
      .map(m -> m.equals(Mailbox.SENT));

      final Observable<RefTag> refTag = intent()
        .flatMap(ProjectIntentMapper::refTag);

      final Observable<Pair<RefTag, KoalaContext.Mailbox>> refTagAndContext = refTag
        .compose(combineLatestPair(koalaContext));

      Observable.combineLatest(projectAndMailbox, refTagAndContext, Pair::create)
        .compose(bindToLifecycle())
        .subscribe(this::trackMailboxView);
    }

    private void trackMailboxView(final @NonNull Pair<Pair<Project, Mailbox>, Pair<RefTag, KoalaContext.Mailbox>> projectMailboxAndRedTag) {
      final Mailbox mailbox = projectMailboxAndRedTag.first.second;
      final Project project = projectMailboxAndRedTag.first.first;
      final RefTag refTag = projectMailboxAndRedTag.second.first;
      final KoalaContext.Mailbox context = projectMailboxAndRedTag.second.second;
      this.koala.trackViewedMailbox(mailbox, project, refTag, context);
    }

    private int getStringResForMailbox(final @NonNull Mailbox mailbox) {
      if (mailbox == Mailbox.INBOX) {
        return R.string.messages_navigation_inbox;
      } else {
        return R.string.messages_navigation_sent;
      }
    }

    private final PublishSubject<Mailbox> mailbox = PublishSubject.create();
    private final PublishSubject<Void> nextPage = PublishSubject.create();
    private final PublishSubject<Void> onResume = PublishSubject.create();
    private final PublishSubject<Void> swipeRefresh = PublishSubject.create();

    private final BehaviorSubject<Boolean> hasNoMessages = BehaviorSubject.create();
    private final BehaviorSubject<Boolean> hasNoUnreadMessages = BehaviorSubject.create();
    private final BehaviorSubject<Boolean> isFetchingMessageThreads = BehaviorSubject.create();
    private final BehaviorSubject<Integer> mailboxTitle = BehaviorSubject.create();
    private final BehaviorSubject<List<MessageThread>> messageThreadList = BehaviorSubject.create();
    private final BehaviorSubject<Integer> unreadCountTextViewColorInt = BehaviorSubject.create();
    private final BehaviorSubject<Integer> unreadCountTextViewTypefaceInt = BehaviorSubject.create();
    private final Observable<Boolean> unreadCountToolbarTextViewIsGone;
    private final BehaviorSubject<Integer> unreadMessagesCount = BehaviorSubject.create();
    private final Observable<Boolean> unreadMessagesCountIsGone;

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void mailbox(final @NonNull Mailbox mailbox) {
      this.mailbox.onNext(mailbox);
    }
    @Override public void nextPage() {
      this.nextPage.onNext(null);
    }
    @Override public void onResume() {
      this.onResume.onNext(null);
    }
    @Override public void swipeRefresh() {
      this.swipeRefresh.onNext(null);
    }

    @Override public @NonNull Observable<Boolean> hasNoMessages() {
      return this.hasNoMessages;
    }
    @Override public @NonNull Observable<Boolean> hasNoUnreadMessages() {
      return this.hasNoUnreadMessages;
    }
    @Override public @NonNull Observable<Boolean> isFetchingMessageThreads() {
      return this.isFetchingMessageThreads;
    }
    @Override public @NonNull Observable<Integer> mailboxTitle() {
      return this.mailboxTitle;
    }
    @Override public @NonNull Observable<List<MessageThread>> messageThreadList() {
      return this.messageThreadList;
    }
    @Override public @NonNull Observable<Integer> unreadCountTextViewColorInt() {
      return this.unreadCountTextViewColorInt;
    }
    @Override public @NonNull Observable<Integer> unreadCountTextViewTypefaceInt() {
      return this.unreadCountTextViewTypefaceInt;
    }
    @Override public @NonNull Observable<Boolean> unreadCountToolbarTextViewIsGone() {
      return this.unreadCountToolbarTextViewIsGone;
    }
    @Override public @NonNull Observable<Integer> unreadMessagesCount() {
      return this.unreadMessagesCount;
    }
    @Override public @NonNull Observable<Boolean> unreadMessagesCountIsGone() {
      return this.unreadMessagesCountIsGone;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/MessageThreadsViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import android.graphics.Typeface
import android.util.Pair
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.R
import com.kickstarter.libs.ApiPaginatorV2
import com.kickstarter.libs.CurrentUserTypeV2
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.intValueOrZero
import com.kickstarter.libs.utils.extensions.isNonZero
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.libs.utils.extensions.isZero
import com.kickstarter.models.MessageThread
import com.kickstarter.models.Project
import com.kickstarter.services.ApiClientTypeV2
import com.kickstarter.services.apiresponses.MessageThreadsEnvelope
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.data.Mailbox
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface MessageThreadsViewModel {
    interface Inputs {
        fun mailbox(mailbox: Mailbox)

        /** Call when pagination should happen.  */
        fun nextPage()

        /** Call when onResume of the activity's lifecycle happens.  */
        fun onResume()

        /** Call when the swipe refresher is invoked.  */
        fun swipeRefresh()
    }

    interface Outputs {
        /** Emits a boolean to determine if there are no messages.  */
        fun hasNoMessages(): Observable<Boolean>

        /** Emits a boolean to determine if there are no unread messages.  */
        fun hasNoUnreadMessages(): Observable<Boolean>

        /** Emits a boolean indicating whether message threads are being fetched from the API.  */
        fun isFetchingMessageThreads(): Observable<Boolean>

        /** Emits a string resource integer to set the mailbox title text view to.  */
        fun mailboxTitle(): Observable<Int>

        /** Emits a list of message threads to be displayed.  */
        fun messageThreadList(): Observable<List<MessageThread>>

        /** Emits a color integer to set the unread count text view to.  */
        fun unreadCountTextViewColorInt(): Observable<Int>

        /** Emits a typeface integer to set the unread count text view to.  */
        fun unreadCountTextViewTypefaceInt(): Observable<Int>

        /** Emits a boolean to determine if the unread count toolbar text view should be gone.  */
        fun unreadCountToolbarTextViewIsGone(): Observable<Boolean>

        /** Emits the unread message count to be displayed.  */
        fun unreadMessagesCount(): Observable<Int>

        /** Emits a boolean determining the unread messages count visibility.  */
        fun unreadMessagesCountIsGone(): Observable<Boolean>
    }

    class MessageThreadsViewModel(environment: Environment, private val intent: Intent? = null) : ViewModel(), Inputs, Outputs {
        private val client: ApiClientTypeV2
        private val currentUser: CurrentUserTypeV2

        private val disposables = CompositeDisposable()
        private fun getStringResForMailbox(mailbox: Mailbox): Int {
            return if (mailbox === Mailbox.INBOX) {
                R.string.messages_navigation_inbox
            } else {
                R.string.messages_navigation_sent
            }
        }

        private val mailbox = PublishSubject.create<Mailbox>()
        private val nextPage = PublishSubject.create<Unit>()
        private val onResume = PublishSubject.create<Unit>()
        private val swipeRefresh = PublishSubject.create<Unit>()
        private val hasNoMessages = BehaviorSubject.create<Boolean>()
        private val hasNoUnreadMessages = BehaviorSubject.create<Boolean>()
        private val isFetchingMessageThreads = BehaviorSubject.create<Boolean>()
        private val mailboxTitle = BehaviorSubject.create<Int>()
        private val messageThreadList = BehaviorSubject.create<List<MessageThread>>()
        private val unreadCountTextViewColorInt = BehaviorSubject.create<Int>()
        private val unreadCountTextViewTypefaceInt = BehaviorSubject.create<Int>()
        private val unreadCountToolbarTextViewIsGone: Observable<Boolean>
        private val unreadMessagesCount = BehaviorSubject.create<Int>()
        private val unreadMessagesCountIsGone: Observable<Boolean>

        val inputs: Inputs = this
        val outputs: Outputs = this

        private fun intent() = intent?.let { Observable.just(it) } ?: Observable.empty()
        init {
            client = requireNotNull(environment.apiClientV2())
            currentUser = requireNotNull(environment.currentUserV2())

            // NB: project from intent can be null.
            val initialProject = intent()
                .distinctUntilChanged()
                .filter { it.getParcelableExtra<Project>(IntentKey.PROJECT) != null }
                .map { i: Intent -> i.getParcelableExtra<Project>(IntentKey.PROJECT) }

            val refreshUserOrProject = Observable.merge(onResume, swipeRefresh)

            intent()
                .filter { it.isNotNull() }
                .compose(Transformers.takeWhenV2(refreshUserOrProject))
                .switchMap {
                    client.fetchCurrentUser()
                }
                .compose(Transformers.neverErrorV2())
                .distinctUntilChanged()
                .subscribe {
                    currentUser.refresh(it)
                }
                .addToDisposable(disposables)

            val refreshedProject = initialProject.compose(Transformers.takeWhenV2(refreshUserOrProject))
                .filter { it.param().isNotNull() }
                .map { it.param() }
                .switchMap { param ->
                    client.fetchProject(param)
                }
                .distinctUntilChanged()
                .compose(Transformers.neverErrorV2())
                .share()

            val project = Observable.merge(
                initialProject,
                refreshedProject
            )

            // vm configured with creator
            val unreadFromUser = currentUser.loggedInUser()
                .map { it.unreadMessagesCount() }

            // Use project unread messages count if configured with a project,
            // in the case of a creator viewing their project's messages
            val unreadFromProject = project
                .map { it.unreadMessagesCount() }

            val unreadMessagesCount =
                Observable.merge(unreadFromUser, unreadFromProject)
                    .distinctUntilChanged()

            // todo: MessageSubject switch will also trigger refresh
            val refreshMessageThreads = Observable.merge(
                unreadMessagesCount.compose(Transformers.ignoreValuesV2()),
                swipeRefresh
            )

            val mailbox = mailbox
                .startWith(Mailbox.INBOX)
                .distinctUntilChanged()

            mailbox
                .map { getStringResForMailbox(it) }
                .subscribe { mailboxTitle.onNext(it) }
                .addToDisposable(disposables)

            val projectAndMailbox = mailbox
                // start wil empty project instead of null
                .withLatestFrom(project.startWith(Project.Builder().build())) { box, proj ->
                    Pair(box, proj)
                }

            val startOverWith =
                Observable.combineLatest(
                    projectAndMailbox,
                    refreshMessageThreads
                ) { a, b: Unit -> a }
                    .distinctUntilChanged()

            val paginator =
                ApiPaginatorV2.builder<MessageThread, MessageThreadsEnvelope, Pair<Mailbox, Project>>()
                    .nextPage(nextPage)
                    .startOverWith(startOverWith)
                    .envelopeToListOfData { it.messageThreads() }
                    .envelopeToMoreUrl {
                        it.urls().api().moreMessageThreads()
                    }
                    .loadWithParams {
                        // - if empty project send null to the API
                        val proj = if (it.second.name().isNotEmpty()) it.second else null
                        client.fetchMessageThreads(
                            proj,
                            it.first
                        )
                    }
                    .loadWithPaginationPath {
                        client.fetchMessageThreadsWithPaginationPath(
                            it
                        )
                    }
                    .clearWhenStartingOver(true)
                    .build()

            paginator.isFetching
                .subscribe { isFetchingMessageThreads.onNext(it) }
                .addToDisposable(disposables)

            paginator.paginatedData()
                .distinctUntilChanged()
                .subscribe {
                    messageThreadList.onNext(it)
                }
                .addToDisposable(disposables)

            unreadMessagesCount
                .map { it.isZero() }
                .subscribe { hasNoMessages.onNext(it) }
                .addToDisposable(disposables)

            unreadMessagesCount
                .map { it.isZero() }
                .subscribe { hasNoUnreadMessages.onNext(it) }
                .addToDisposable(disposables)

            unreadMessagesCount
                .filter { it.isNotNull() }
                .map { requireNotNull(it) }
                .map { if (it.intValueOrZero() > 0) R.color.accent else R.color.kds_support_400 }
                .subscribe { unreadCountTextViewColorInt.onNext(it) }
                .addToDisposable(disposables)

            unreadMessagesCount
                .filter { it.isNotNull() }
                .map { requireNotNull(it) }
                .map { if (it.intValueOrZero() > 0) Typeface.BOLD else Typeface.NORMAL }
                .subscribe { unreadCountTextViewTypefaceInt.onNext(it) }
                .addToDisposable(disposables)

            unreadCountToolbarTextViewIsGone =
                Observable.zip<Boolean, Boolean, Pair<Boolean, Boolean>>(
                    hasNoMessages,
                    hasNoUnreadMessages
                ) { a: Boolean?, b: Boolean? -> Pair.create(a, b) }
                    .map { noMessagesAndNoUnread: Pair<Boolean, Boolean> -> noMessagesAndNoUnread.first || noMessagesAndNoUnread.second }
                    .compose(Transformers.combineLatestPair(mailbox))
                    .map { noMessagesAndMailbox: Pair<Boolean, Mailbox> -> noMessagesAndMailbox.first || noMessagesAndMailbox.second == Mailbox.SENT }

            unreadMessagesCount
                .filter { it.isNotNull() }
                .filter { it.isNonZero() }
                .subscribe(this.unreadMessagesCount)
            unreadMessagesCountIsGone = mailbox
                .map { m: Mailbox -> m == Mailbox.SENT }
        }
        override fun mailbox(mailbox: Mailbox) {
            this.mailbox.onNext(mailbox)
        }

        override fun nextPage() {
            nextPage.onNext(Unit)
        }

        override fun onResume() {
            onResume.onNext(Unit)
        }

        override fun swipeRefresh() {
            swipeRefresh.onNext(Unit)
        }

        override fun onCleared() {
            disposables.clear()
            super.onCleared()
        }

        override fun hasNoMessages(): Observable<Boolean> = hasNoMessages
        override fun hasNoUnreadMessages(): Observable<Boolean> = hasNoUnreadMessages
        override fun isFetchingMessageThreads(): Observable<Boolean> = isFetchingMessageThreads
        override fun mailboxTitle(): Observable<Int> = mailboxTitle
        override fun messageThreadList(): Observable<List<MessageThread>> = messageThreadList
        override fun unreadCountTextViewColorInt(): Observable<Int> = unreadCountTextViewColorInt
        override fun unreadCountTextViewTypefaceInt(): Observable<Int> = unreadCountTextViewTypefaceInt
        override fun unreadCountToolbarTextViewIsGone(): Observable<Boolean> = unreadCountToolbarTextViewIsGone
        override fun unreadMessagesCount(): Observable<Int> = unreadMessagesCount
        override fun unreadMessagesCountIsGone(): Observable<Boolean> = unreadMessagesCountIsGone
    }

    class Factory(private val environment: Environment, private val intent: Intent) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return MessageThreadsViewModel(environment, intent) as T
        }
    }
}
